src/main/java/de/tuclausthal/submissioninterface/servlets/controller/HaskellRuntimeTestManager.java

```diff
@@ -0,0 +1,1116 @@
+ * Controller-Servlet for clustering haskell submissions based on common errors (dynamic/runtime analysis).
+ * This servlet allows advisors to automatically generate and modify test steps.
+ *
+ * @author Christian Wagner
+ */
+@GATEController
+public class HaskellRuntimeTestManager extends HttpServlet {
+	@Serial
+	private static final long serialVersionUID = 1L;
+	final static private Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+	// TODO@CHW: should safe docker path be part of de.tuclausthal.submissioninterface.util.Configuration? (since it is duplicated from DockerTest)
+	final static private String SAFE_DOCKER_SCRIPT = "/usr/local/bin/safe-docker";
+
+	@Override
+	public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+		getServletContext().getNamedDispatcher(DockerTestManager.class.getSimpleName()).forward(request, response);
+	}
+
+	@Override
+	public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+		Session session = RequestAdapter.getSession(request);
+		TestDAOIf testDAOIf = DAOFactory.TestDAOIf(session);
+		Test test = testDAOIf.getTest(Util.parseInteger(request.getParameter("testid"), 0));
+		if (!(test instanceof HaskellRuntimeTest haskellRuntimeTest)) {
+			response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+			request.setAttribute("title", "Test nicht gefunden");
+			getServletContext().getNamedDispatcher(MessageView.class.getSimpleName()).forward(request, response);
+			return;
+		}
+
+		ParticipationDAOIf participationDAO = DAOFactory.ParticipationDAOIf(session);
+		Participation participation = participationDAO.getParticipation(RequestAdapter.getUser(request), haskellRuntimeTest.getTask().getTaskGroup().getLecture());
+		if (participation == null || participation.getRoleType() != ParticipationRole.ADVISOR) {
+			response.sendError(HttpServletResponse.SC_FORBIDDEN, "insufficient rights");
+			return;
+		}
+
+		if ("browseModelSolution".equals(request.getParameter("action"))) {
+			deleteStoredClassifiedIdentifiers(haskellRuntimeTest, session);
+			try {
+				browseModelSolutionAndStoreClassifiedIdentifiers(haskellRuntimeTest, session);
+			} catch (IOException e) {
+				request.getSession().setAttribute("haskellRuntimeTestBrowseError", e.getMessage());
+			}
+			response.sendRedirect(Util.generateRedirectURL(HaskellRuntimeTestManager.class.getSimpleName() + "?testid=" + haskellRuntimeTest.getId(), response));
+		} else if ("deleteHaskellIdentifiers".equals(request.getParameter("action"))) {
+			deleteStoredClassifiedIdentifiers(haskellRuntimeTest, session);
+			response.sendRedirect(Util.generateRedirectURL(HaskellRuntimeTestManager.class.getSimpleName() + "?testid=" + haskellRuntimeTest.getId(), response));
+		} else if ("generateFunctionTestcases".equals(request.getParameter("action"))) {
+			int identifierId = Util.parseInteger(request.getParameter("identifierid"), -1);
```

Comment:
Is it ensured that the user-entered `identifierId` belongs to the test resp. cannot refer to another test for which we do not have permissions?